### ISSUE TITLE
Closure example

### ISSUE THREAD

**Issue by BogdanVU** [chesdenis/net-courses#78]
@chesdenis Напомните, пожалуйста, ссылку с примером про замыкание.

**Comment by tyrgin** [comment]
А можете, пожалуйста, пояснить, почему замыкание это фича, а не баг? В чем его сила?)

**Comment by chesdenis** [comment]
http://aakinshin.net/ru/blog/post/closures/

**Comment by tyrgin** [comment]
@chesdenis  да, я это читал, но как по мне - это пояснения механизма. Увидеть бы пример "обычно делаем вот так, а с захватом получается лучше, вот так"

**Comment by backsapc** [comment]
Очень странный вопрос, представь, что для использования какой-то переменной внутри лямбда выражения нужно было бы явно передавать её в его область видимости. У тебя загрязняются аргументы функции, код выглядит менее читабельным, банально будет сложно сопоставить сигнатуры целевого и твоего метода. А так тебя избавили от лишних проблем (и добавили новых).

**Comment by BogdanVU** [comment]
`class MyClass
    {
        public int i;
        ~MyClass()
        {
            Console.WriteLine("Destructor MyClass");
        }
    }

    class MyClassViaDisplayClass
    {
        public int i;
        ~MyClassViaDisplayClass()
        {
            Console.WriteLine("Destructor MyClassViaDisplayClass");
        }
    }

    sealed class MyDisplayClass
    {
        public MyClassViaDisplayClass mc;

        public void Action()
        {
            mc = null;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Explisit Closure");
            var actions1 = new List<Action>();
            MyClassViaDisplayClass mc1 = new MyClassViaDisplayClass() { i = 11 };
            MyDisplayClass c = new MyDisplayClass() { mc = mc1 };
            for (int i = 0; i < 3; i++)
                actions1.Add(c.Action);
            foreach (Action action in actions1)
                action();
            Console.WriteLine(mc1.i);

            Console.WriteLine("Implicit Closure");
            var actions = new List<Action>();
            MyClass mc = new MyClass() { i = 11 };
            for (int i = 0; i < 3; i++)
                actions.Add(() => { Console.WriteLine("Actoin!"); mc = null; });
            Console.WriteLine(mc.i);
            foreach (var action in actions)
                action();
            Console.WriteLine(mc.i);
            
        }
    }`

@chesdenis Почему в этом случае если мы явно сами уходим от замыкания и пишем свой MyDisplayClass, то mc1 не обнуляется в main, а в "неявном" замыкании mc обнуляется? Ведь я развернул этот MyDisplayClass ровно так, как указано в примере из ссылки. Понятно почему mc1 не обнуляется в Main, но значит тогда то, что указано в статье неверно? Или я чего-то не так развернул или не понимаю?

**Comment by BogdanVU** [comment]
И еще вопрос.
`class MyClass
    {
        public int i;
        public MyClass(int i)
        {
            this.i = i;
            Console.WriteLine($"Instance {i} has been created.");
        }
        ~MyClass()
        {
            Console.WriteLine($"Instance {i} has been disposed.");
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            for (int i=0;i<3;i++)
            {
                MyClass mc = new MyClass(i);
            }
            GC.Collect();
            Console.ReadKey(true);
        }
    }
`

В результате будет выведено примерно следующее:

Instance 0 has been created.
Instance 1 has been created.
Instance 2 has been created.
Instance 1 has been disposed.
Instance 0 has been disposed.

Почему не уничтожается Instance 2?

**Comment by backsapc** [comment]
По твоему первому вопросу, ты не совсем условил суть этого всего. После того, как компилятор собственно произвёл компиляцию твоего кода, у него действительно получилось похожее определение класса, как и у тебя. 
![image](https://user-images.githubusercontent.com/18036840/39078351-d4d11360-4511-11e8-94d6-3e7d154244a8.png)
Но в этом месте ты написал не совсем так, твоей переменной mc1 как бы больше не существует, 
эквивалентным кодом для тебя было бы что-то вроде 
MyDisplayClass c = new MyDisplayClass() { mc = new MyClassViaDisplayClass() { i = 11 } };
![image](https://user-images.githubusercontent.com/18036840/39078373-18f0e598-4512-11e8-858c-eaf5f5668eb9.png)
Здесь же в свою очередь должно быть что-то вроде 
Console.WriteLine(c.mc.i); // Бросит NullReferenceException
Если нормально с английским, то вот https://blogs.msdn.microsoft.com/abhinaba/2005/10/18/c-anonymous-methods-are-not-closures/
